File: apps/photokit/photokit/fix_dates.py
#!/usr/bin/env python3
"""
Restore EXIF date metadata from filenames in a photo directory.

Can be run stand-alone:

    python -m photokit.fix_dates /path/to/photos --dry-run

Or invoked as ``photokit fix-dates`` via the Click CLI.
"""

import argparse
import sys
from pathlib import Path
from typing import Optional

import piexif
from PIL import Image

from .date_patterns import (
    auto_detect,
    match_from_pattern,
    DATE_PATTERN_HELP,
)

IMAGE_EXTENSIONS = {".jpg", ".jpeg", ".png", ".tiff", ".tif", ".heic", ".heif", ".webp"}


def find_images(directory: Path, extensions: set, recursive: bool) -> list[Path]:
    """Find image files matching *extensions* under *directory*."""
    if recursive:
        files = [p for ext in extensions for p in directory.rglob(f"*{ext}")]
    else:
        files = [p for ext in extensions for p in directory.glob(f"*{ext}")]
    # Filter out symlinks, macOS extended attributes, etc.
    return sorted(
        p for p in files
        if p.is_file() and not p.is_symlink() and not p.name.startswith("._")
    )


def create_exif_dict(
    year: int, month: int, day: int,
    hour: int, minute: int, second: int,
) -> dict:
    """Build an EXIF dict with date/time tags."""
    date_str = f"{year:04d}:{month:02d}:{day:02d} {hour:02d}:{minute:02d}:{second:02d}"
    return {
        "0th": {piexif.ImageIFD.DateTime: date_str},
        "Exif": {
            piexif.ExifIFD.DateTimeOriginal: date_str,
            piexif.ExifIFD.DateTimeDigitized: date_str,
        },
    }


def write_exif_date(filepath: Path, date_parts: dict) -> bool:
    """Write EXIF date metadata into an image file in-place.

    Uses ``piexif.insert`` which works for both JPEG and PNG (inserts an
    eXIf chunk into PNGs — no format conversion needed).
    """
    try:
        exif_dict = create_exif_dict(**date_parts)
        exif_bytes = piexif.dump(exif_dict)
        # piexif.insert writes directly into the binary — works for
        # JPEG and PNG without re-encoding.
        piexif.insert(exif_bytes, str(filepath))
        return True
    except Exception as exc:
        print(f"  ❌ Error writing EXIF to {filepath.name}: {exc}", file=sys.stderr)
        return False


def fix_dates(
    directory: Path,
    *,
    pattern: Optional[str] = None,
    recursive: bool = False,
    dry_run: bool = False,
    extensions: Optional[set] = None,
) -> int:
    """Main logic for fix-dates.

    Returns exit code (0 = success).
    """
    if extensions is None:
        extensions = IMAGE_EXTENSIONS

    images = find_images(directory, extensions, recursive)
    if not images:
        print(f"No images found in {directory}")
        return 0

    print(f"\n📷  Found {len(images)} image(s) in {directory}\n")

    fixed = 0
    skipped = 0
    errors = 0

    for filepath in images:
        if pattern:
            date_parts = match_from_pattern(pattern, filepath.name)
        else:
            date_parts = auto_detect(filepath.name)

        if date_parts is None:
            print(f"  ⚠️  Could not parse date from: {filepath.name}")
            skipped += 1
            continue

        ds = date_parts
        print(
            f"  📷 {filepath.name}\n"
            f"     Extracted date: {ds['year']:04d}-{ds['month']:02d}-{ds['day']:02d}"
            f" {ds['hour']:02d}:{ds['minute']:02d}:{ds['second']:02d}"
        )

        if dry_run:
            print("     (dry run — skipped)")
            fixed += 1  # count as "would fix"
            continue

        if write_exif_date(filepath, date_parts):
            fixed += 1
        else:
            errors += 1

    print(f"\n{'=' * 50}")
    print(f"  Fixed:    {fixed}")
    print(f"  Skipped:  {skipped}")
    print(f"  Errors:   {errors}")
    return 1 if errors else 0


# ---------------------------------------------------------------------------
# Stand-alone entry point
# ---------------------------------------------------------------------------

def main(argv: Optional[list[str]] = None) -> int:
    parser = argparse.ArgumentParser(
        description="Restore EXIF date metadata from filenames in a photo directory.",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=DATE_PATTERN_HELP,
    )
    parser.add_argument("directory", type=Path, help="Directory containing photos")
    parser.add_argument(
        "-p", "--pattern",
        help="Custom regex with named groups year, month, day (or year, doy)",
    )
    parser.add_argument(
        "-r", "--recursive", action="store_true",
        help="Recurse into subdirectories",
    )
    parser.add_argument(
        "-n", "--dry-run", action="store_true",
        help="Preview only — don't write EXIF data",
    )
    parser.add_argument(
        "--extensions",
        default=",".join(sorted(IMAGE_EXTENSIONS)),
        help=f"Comma-separated extensions (default: {','.join(sorted(IMAGE_EXTENSIONS))})",
    )

    args = parser.parse_args(argv)
    extensions = {f".{e.strip().lstrip('.')}" for e in args.extensions.split(",")}

    return fix_dates(
        args.directory,
        pattern=args.pattern,
        recursive=args.recursive,
        dry_run=args.dry_run,
        extensions=extensions,
    )


if __name__ == "__main__":
    sys.exit(main())

File: apps/imagekit/src/commands/fix-dates.ts
import fs from "node:fs";
import path from "node:path";
import { run, checkCmd } from "../utils";
import { autoDetect, matchFromPattern } from "../date-patterns";

interface FixDatesOptions {
  pattern?: string;
  recursive: boolean;
  dryRun: boolean;
  extensions?: string;
}

const DEFAULT_EXTENSIONS = new Set([
  ".jpg",
  ".jpeg",
  ".png",
  ".tiff",
  ".tif",
  ".heic",
  ".heif",
  ".webp",
]);

function parseExtensions(extStr?: string): Set<string> {
  if (!extStr) return DEFAULT_EXTENSIONS;
  const set = new Set<string>();
  for (const e of extStr.split(",")) {
    const trimmed = e.trim().toLowerCase();
    if (!trimmed) continue;
    const withDot = trimmed.startsWith(".") ? trimmed : `.${trimmed}`;
    set.add(withDot);
  }
  return set.size > 0 ? set : DEFAULT_EXTENSIONS;
}

function findImages(directory: string, extensions: Set<string>, recursive: boolean): string[] {
  const results: string[] = [];
  const dir = path.resolve(directory);

  function walk(current: string) {
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(current, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of entries) {
      if (entry.name.startsWith("._")) continue;
      const full = path.join(current, entry.name);
      if (entry.isSymbolicLink()) continue;
      if (entry.isDirectory()) {
        if (recursive) walk(full);
      } else if (entry.isFile()) {
        const ext = path.extname(entry.name).toLowerCase();
        if (extensions.has(ext)) results.push(full);
      }
    }
  }

  walk(dir);
  return results.sort();
}

function createExifDateString(parts: { year: number; month: number; day: number; hour: number; minute: number; second: number }): string {
  return `${String(parts.year).padStart(4, "0")}:${String(parts.month).padStart(2, "0")}:${String(parts.day).padStart(2, "0")} ${String(parts.hour).padStart(2, "0")}:${String(parts.minute).padStart(2, "0")}:${String(parts.second).padStart(2, "0")}`;
}

function writeExifDate(filepath: string, dateParts: { year: number; month: number; day: number; hour: number; minute: number; second: number }): boolean {
  const dateStr = createExifDateString(dateParts);
  // Use exiftool to write; -overwrite_original to avoid backup
  // Write to common date tags: DateTimeOriginal, CreateDate, ModifyDate
  // Also FileModifyDate via exiftool?
  // For PNG, exiftool will add appropriate chunks.
  const args = [
    "exiftool",
    "-overwrite_original",
    `-DateTimeOriginal=${dateStr}`,
    `-CreateDate=${dateStr}`,
    `-ModifyDate=${dateStr}`,
    `-FileModifyDate=${dateStr}`,
    filepath,
  ];
  const r = run(args);
  if (r.exitCode !== 0) {
    console.error(`  ❌ Error writing EXIF to ${path.basename(filepath)}: ${r.stderr.trim() || r.stdout.trim()}`);
    return false;
  }
  return true;
}

export function cmdFixDates(directory: string, opts: FixDatesOptions): void {
  checkCmd("exiftool", "brew install exiftool");

  const extensions = parseExtensions(opts.extensions);
  const dir = directory ?? ".";
  if (!fs.existsSync(dir) || !fs.statSync(dir).isDirectory()) {
    console.error(`Error: directory not found: ${dir}`);
    process.exit(1);
  }

  const images = findImages(dir, extensions, !!opts.recursive);
  if (images.length === 0) {
    console.log(`No images found in ${dir}`);
    return;
  }

  console.log(`\n📷  Found ${images.length} image(s) in ${dir}\n`);

  let fixed = 0;
  let skipped = 0;
  let errors = 0;

  for (const filepath of images) {
    const name = path.basename(filepath);
    let dateParts: ReturnType<typeof autoDetect> = null;
    try {
      if (opts.pattern) {
        dateParts = matchFromPattern(opts.pattern, name);
      } else {
        dateParts = autoDetect(name);
      }
    } catch (e) {
      console.error(`  ❌ Invalid pattern: ${e instanceof Error ? e.message : String(e)}`);
      process.exit(1);
    }

    if (!dateParts) {
      console.log(`  ⚠️  Could not parse date from: ${name}`);
      skipped++;
      continue;
    }

    const ds = dateParts;
    console.log(
      `  📷 ${name}\n     Extracted date: ${String(ds.year).padStart(4, "0")}-${String(ds.month).padStart(2, "0")}-${String(ds.day).padStart(2, "0")} ${String(ds.hour).padStart(2, "0")}:${String(ds.minute).padStart(2, "0")}:${String(ds.second).padStart(2, "0")}`,
    );

    if (opts.dryRun) {
      console.log("     (dry run — skipped)");
      fixed++;
      continue;
    }

    if (writeExifDate(filepath, dateParts)) {
      fixed++;
    } else {
      errors++;
    }
  }

  console.log(`\n${"=".repeat(50)}`);
  console.log(`  Fixed:    ${fixed}`);
  console.log(`  Skipped:  ${skipped}`);
  console.log(`  Errors:   ${errors}`);
  if (errors > 0) process.exitCode = 1;
}
